File: anb/leetcode/graph/207. Course Schedule.cpp
const int MAX_N = 1e5;
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int ind[MAX_N]={}, i, a, b;
        bool ans = true;
        queue<int> q;
        vector<int> adj[MAX_N];
        for(vector<int>& e : prerequisites) {
            a = e[0], b = e[1];
            adj[b].push_back(a);
            ind[a]++;
        }
        for(i=0;i<numCourses;i++)
            if(ind[i] == 0)
                q.push(i);
        while(!q.empty()){
            a = q.front(); q.pop();
            for(int& v : adj[a]) {
                ind[v] -= 1;
                if(ind[v] == 0)
                    q.push(v);
            }
        }
        
        for(i=0;i<numCourses;i++)
            ans &= ind[i] == 0;
        return ans;
    }
};

File: gunkim/leetcode/graph/207.Course Schedule.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.canFinish(5, new int[][] { { 1, 2 } }));
    }
}

class Solution {
    Map<Integer, ArrayList<Integer>> adj = new HashMap<>();
    int[] deg;

    boolean bfs() {
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < deg.length; i++) {
            if (deg[i] == 0) {
                q.add(i);
            }
        }
        while (q.isEmpty() == false) {
            int curr = q.poll();
            ArrayList<Integer> children = adj.get(curr);
            if (children != null) {
                for (int child : children) {
                    deg[child]--;
                    if (deg[child] == 0)
                        q.add(child);
                }
            }

        }
        for (int i = 0; i < deg.length; i++) {
            if (deg[i] != 0)
                return false;
        }
        return true;
    }

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int n = prerequisites.length;
        deg = new int[numCourses];
        for (int i = 0; i < n; i++) {
            int a = prerequisites[i][0];
            int b = prerequisites[i][1];
            if (!adj.containsKey(b)) {
                adj.put(b, new ArrayList<Integer>());
            }
            adj.get(b).add(a);
            deg[a]++;
        }
        return bfs();

    }
}